File: 3. Behavioral Design Patterns/Strategy/Go/main.go
package main

import "fmt"

// Strategy Pattern Example in Go
// Sorter with BubbleSort and QuickSort strategies

type SortStrategy interface{ Sort([]int) []int }

type BubbleSort struct{}
func (BubbleSort) Sort(data []int) []int {
	arr := append([]int(nil), data...)
	for i := 0; i < len(arr); i++ {
		for j := 0; j < len(arr)-i-1; j++ {
			if arr[j] > arr[j+1] { arr[j], arr[j+1] = arr[j+1], arr[j] }
		}
	}
	fmt.Println("BubbleSort used")
	return arr
}

type QuickSort struct{}
func (QuickSort) Sort(data []int) []int {
	arr := append([]int(nil), data...)
	// simple selection sort for demo brevity
	for i := 0; i < len(arr); i++ {
		minIdx := i
		for j := i+1; j < len(arr); j++ { if arr[j] < arr[minIdx] { minIdx = j } }
		arr[i], arr[minIdx] = arr[minIdx], arr[i]
	}
	fmt.Println("QuickSort used")
	return arr
}

type Sorter struct{ s SortStrategy }
func (so *Sorter) SetStrategy(s SortStrategy){ so.s = s }
func (so Sorter) Sort(data []int) []int { return so.s.Sort(data) }

func main(){
	data := []int{5,2,9,1,5,6}
	so := Sorter{ s: BubbleSort{} }
	fmt.Println(so.Sort(data))
	so.SetStrategy(QuickSort{})
	fmt.Println(so.Sort(data))
}

File: 3. Behavioral Design Patterns/Strategy/Rust/main.rs
// Strategy Pattern Example in Rust
// Sorter with BubbleSort and QuickSort strategies

trait SortStrategy {
    fn sort(&self, data: &[i32]) -> Vec<i32>;
}

struct BubbleSort;
impl SortStrategy for BubbleSort {
    fn sort(&self, data: &[i32]) -> Vec<i32> {
        let mut arr = data.to_vec();
        for i in 0..arr.len() {
            for j in 0..arr.len().saturating_sub(i + 1) {
                if arr[j] > arr[j + 1] {
                    arr.swap(j, j + 1);
                }
            }
        }
        println!("BubbleSort used");
        arr
    }
}

struct QuickSort;
impl SortStrategy for QuickSort {
    fn sort(&self, data: &[i32]) -> Vec<i32> {
        let mut arr = data.to_vec();
        arr.sort(); // brevity
        println!("QuickSort used");
        arr
    }
}

struct Sorter<T: SortStrategy> {
    strategy: T,
}

impl<T: SortStrategy> Sorter<T> {
    fn new(strategy: T) -> Self { Self { strategy } }
    fn with_strategy<U: SortStrategy>(self, strategy: U) -> Sorter<U> { Sorter { strategy } }
    fn sort(&self, data: &[i32]) -> Vec<i32> { self.strategy.sort(data) }
}

fn main() {
    let data = vec![5, 2, 9, 1, 5, 6];
    let sorter = Sorter::new(BubbleSort);
    println!("{:?}", sorter.sort(&data));
    let sorter = sorter.with_strategy(QuickSort);
    println!("{:?}", sorter.sort(&data));
}
